### PR TITLE
Short-circuit to `dsl.instance_exec` when `dsl.equal?(block_context)`

### PR DESCRIPTION
**Pull request by tomeon**
I ran into an issue invoking `Docile.dsl_eval_with_block_return` using a frozen object as the DSL
object, even though none of the code in the evaluated block attempts to mutate the frozen object:

```ruby
# do_something.rb

require 'docile'

class MyClass
  def initialize
    @something = proc { puts "Hello!" }
  end

  def evaluate(*args, &block)
    Docile.dsl_eval_with_block_return(self, *args, &block)
  end

  def something
    @something = Proc.new if block_given?
    @something
  end

  def do_something
    evaluate(&something)
  end
end

MyClass.new.do_something
MyClass.new.tap { |mc| mc.something { puts "Goodbye!" } }.freeze.do_something
MyClass.new.freeze.do_something
```

```
$ ruby ./do_something.rb
Hello!
Goodbye!
Traceback (most recent call last):
        9: from -e:1:in `<main>'
        8: from -e:1:in `load'
        7: from /path/to/do_something.rb:26:in `<top (required)>'
        6: from /path/to/do_something.rb:20:in `do_something'
        5: from /path/to/do_something.rb:11:in `evaluate'
        4: from /usr/lib/ruby/gems/2.5.0/gems/docile-1.3.2/lib/docile.rb:83:in `dsl_eval_with_block_return'
        3: from /usr/lib/ruby/gems/2.5.0/gems/docile-1.3.2/lib/docile/execution.rb:19:in `exec_in_proxy_context'
        2: from /usr/lib/ruby/gems/2.5.0/gems/docile-1.3.2/lib/docile/execution.rb:19:in `new'
        1: from /usr/lib/ruby/gems/2.5.0/gems/docile-1.3.2/lib/docile/fallback_context_proxy.rb:55:in `initialize'
/nix/store/rlgw8kj33z52s53crdsjdb83wrh63550-ruby2.5.7-docile-1.3.2/lib/ruby/gems/2.5.0/gems/docile-1.3.2/lib/docile/fallback_context_proxy.rb:55:in `define_method': can't modify frozen object (FrozenError)
```

The proximate cause of the error appears to be calling `#define_method` on the frozen object.

Given that, in the scenario that results in a `FrozenError`, the DSL object and the block context
object are identical, it looks like it should be possible to short-circuit to calling
`#instance_exec` on the DSL object rather than setting up a proxy object.  This PR implements
exactly this short-circuit.

Thanks in advance for your consideration!

### CODE REVIEW

**Comment by ms-ati** [comment]
Hi @tomeon thanks for the bug report and the proposed fix PR! Apologies for the delay, I will get back to you with more thoughts soon.

I'm willing accept a special case for `dsl` is the same object as `block_context`, but it would NOT be my first choice when I think about this issue. A test case would be a DSL block that tries to call a method from outside the block's context. It appears that even with your fix, this would still be broken if the dsl object was frozen?

Is it possible that a more encompassing solution would be to find a way to do the method callbacks to other contexts, that works even if the target DSL object is frozen. But if this is not possible or very very difficult, I suppose your solution might be the best that can be easily done?

**Comment by tomeon** [comment]
Hi @ms-ati -- thank you for your reply!

> I'm willing accept a special case for dsl is the same object as block_context, but it would NOT be my first choice when I think about this issue.

My instinct was that it makes sense to short-circuit to `#instance_exec` when the DSL and block context objects are identical, quite apart from the exception I hit, as it cuts out no-ops like setting instance variables to the same values they already contain.  Happily, this optimization incidentally solves my `FrozenError`, too :)

> A test case would be a DSL block that tries to call a method from outside the block's context. It appears that even with your fix, this would still be broken if the dsl object was frozen?

With the short-circuiting change, and assuming that the DSL block does not call code that attempts to mutate the DSL object, I believe the matrix breaks down as follows:

| DSL & block context identical | DSL frozen | Block context frozen | Raises {Frozen,Runtime}Error |
| ------------------------------------ | -------------- | -------------------------- | -------------------------------------- |
| Yes                                        | Yes             | Yes                            | No                                            |
| Yes                                        | No              | No                              | No                                            |
| No                                         | Yes             | Yes                             | **Yes**                                      |
| No                                         | Yes             | No                              | No                                            |
| No                                         | No              | Yes                             | **Yes**                                      |
| No                                         | No              | No                              | No                                            |

Have I understood your concern correctly?  I'm not sure I grok "a DSL block that tries to call a method from outside the block's context"; might you have (pseudo-)code that illustrates this scenario?

> Is it possible that a more encompassing solution would be to find a way to do the method callbacks to other contexts, that works even if the target DSL object is frozen. But if this is not possible or very very difficult, I suppose your solution might be the best that can be easily done?

One (suspicously-too-)easy approach that appears to work is:

```ruby
block_context = eval("self", block.binding)
return dsl.instance_exec(*args, &block) if dsl.equal?(block_context)
block_context = block_context.dup if block_context.frozen?
```

On the one hand, this is a breaking change.  On the other, I suspect it breaks only an arguably already-broken scenario.  Maybe Docile could also issue a warning message if the block context object is frozen?

**Comment by ms-ati** [comment]
@tomeon Thanks for your deep thought and engagement on this. Apologies for the delay, I'm out of town with family. 

Can you share your thinking on the following question: _Why use Docile at all in the short-circuit case?_

That is -- the problem space of Docile, as I conceived of it, was to solve the "harder" problems that a bare call to `instance_exec` cannot solve. The canonical cases are to refer to _variables, instance variables, and methods_ which are defined in the context where the block is written, but are not available on the DSL object itself.

```ruby
# context where programming is being done
a_variable = 1
@an_instance_variable = 2

def a_method(a, b)
  a + b
end

dsl = []

Docile.dsl_eval_with_block_return(dsl) do
  # context of block where Array's methods form a DSL
  append a_method(a_variable, @an_instance_variable)
end

#=> [3]
```

If the goal of the usage is to short-circuit all the code that makes this possible, I'd like to understand which value the usage of Docile is adding? Very happy to learn if I'm missing it!

**Comment by tomeon** [comment]
> Why use Docile at all in the short-circuit case?

DRYness, largely: I hit the `FrozenError` while working on a library that performs certain data transformations that can be customized via caller-provided closures.  In the absence of a caller-provided closure for a given transformation, the library itself defines a default closure.  It's in this latter case that the `FrozenError` occurs, because such default closures are defined within the scope of the same object used as the DSL context for `Docile.dsl_eval_with_block_return`.

Docile is _absolutely essential_ for sensibly evaluating code of the first kind (caller-provided), but not for the second kind.  I can accept that DRYness may not be compelling enough to add the short-circuiting logic to Docile itself; in fact, I've already implemented the short-circuit in my own library.  However, I think it's worth noting that my library isn't the only one using Docile for evaluating internal code -- for instance, [SimpleCov does it too](https://github.com/colszowka/simplecov/blob/5bbad8d2dff8c45a0d3246837c44242b53cdf774/lib/simplecov/defaults.rb) (`SimpleCov.configure` uses `Docile.dsl_eval`), though admittedly in a manner that's not susceptible to the `FrozenError` I hit.

Here are some other considerations that argue in favor of incorporating the short-circuiting logic into Docile, or at least in favor of the proposition that its incorporation would be innocuous:

1. The changeset is small: the meat of the change is a single line of code, though of course there's some additional maintenance burden from the associated test case.
2. The changeset is low-risk: what Docile accomplishes (handily!) is merging a subset of the lexical environment where a block is **defined** with the lexical environment where it is **evaluated**.  The (object-)identity of the DSL object and block context guarantees that this merging, in effect, _already holds_.
3. The `FrozenError` is surprising: although Docile is, as you said, intended 'to solve the "harder" problems that a bare call to `instance_exec` cannot solve', nothing I've come across in Docile's documentation _discourages_ using `dsl_eval` and friends where a plain old `instance_exec` would suffice.  Additionally, at present Docile neither prohibits nor warns about frozen block context objects, so it came as a surprise to me that evaluating (ostensibly) non-mutative code resulted in a `FrozenError`.  
4. Averting the surprise mentioned above is about as expensive as obviating it: updating the docs to raise  caveats about using `dsl_eval` when the the DSL object and block context are identical, etc., requires scarcely less work than simply eliminating the reason why such warnings would be appropriate.  Such caveats are also more susceptible to errors, as Docile's users would have to both read and heed them.

I realize that I'm asking you to take on a maintenance burden (and perhaps a source of bugs) in order to make my code somewhat cleaner.  That, by itself, is not a knock-down value proposition :grimacing: .  But, if you buy the arguments above, I'd be most obliged if you'd hit the ol' "merge" button :grin:

Thanks again for your consideration!

File: spec/docile_spec.rb [hunk 1]
```diff
@@ -414,6 +414,12 @@ def dsl_eval_string(string)
         expect(dsl.foo).to eq(0)
         expect(dsl.bar).to eq(1)
       end
+
+      context "when the DSL object is frozen" do
+        it "can call non-mutative code without raising an exception" do
+          expect { dsl.freeze.dsl_eval_string('1 + 2') }.not_to raise_error
```
Comment (taichi-ishitani) [inline spec/docile_spec.rb]: Hi @tomeon ,
I think the `dsl` object is not frozen so you need to freeze the `dsl` object before evaluating the example which you added.

**Comment by ms-ati** [comment]
@tomeon the bad news is it's been a year -- the good news is that I'd be happy to merge this.

If you are still around and interested, could I persuade you to perhaps rebase this PR on the latest master?

**Comment by tomeon** [comment]
Hi @ms-ati -- thank you very much for merging this!  I'm sorry that I did not respond promptly to your message; I meant to post here about having done the rebase, but got distracted :P

**Pull request merged**